File: LeetCode/101SymmetricTree/SymmetricTree.py
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # DFS -> Time:O(n), Space:O(h)
        def dfs(left, right):
            if not left and not right:
                return True
            elif not left or not right:
                return False
            return left.val == right.val and dfs(left.left, right.right) and dfs(left.right, right.left)
        if root == None:
            return True
        return dfs(root.left, root.right)

File: LeetCode/101SymmetricTree/SymmetricTree.cpp
class Solution {
public:
    bool dfs(TreeNode* left, TreeNode* right) {
        if (left == NULL && right == NULL) {
            return true;
        }
        else if (left == NULL || right == NULL) {
            return false;
        }
        return left->val == right->val && dfs(left->left, right->right) && dfs(left->right, right->left);
    }
    bool isSymmetric(TreeNode* root) {
        // DFS -> Time:O(n), Space:O(h)
        if (root == NULL) {
            return true;
        }
        return dfs(root->left, root->right);
    }
};
